dask_searchcv/hyperband/algs.py

```diff
@@ -0,0 +1,109 @@
+import math
+import numpy as np
+import scipy.stats as stats
+
+import sklearn
+from sklearn.model_selection import train_test_split
+from sklearn import clone
+
+
+def train(model, X, y, X_val, y_val, max_iter=1, classes=None):
+    for iter in range(int(max_iter)):
+        _ = model.partial_fit(X, y, classes)
+        model._iters += 1
+    return model
+
+def predict_and_loss(model, X, y):
+    y_hat = model.predict(X)
+    return model.loss_fn(y_hat, y)
+
+def bottom_k(choose_from, eval_with, k=1):
+    assert k > 0, "k must be positive"
+    assert set(choose_from.keys()) == set(eval_with.keys())
+
+    keys = list(choose_from.keys())
+
+    eval_with = [eval_with[key] for key in keys]
+    idx = np.argsort(eval_with)
+
+    return {keys[i]: choose_from[keys[i]] for i in idx[:k]}
+
+
+def random_choice(params):
+    choice = {}
+    for k, v in params.items():
+        dist = stats.uniform(v['min'], v['max'] - v['min'])
+        r = dist.rvs()
+        if v.get('type', 'linear') == 'log':
+            value = 10**r
+        choice[k] = value
+    return choice
+
+
+class Hyperband:
+    def __init__(self, params, model, R, X, y, eta=3):
+        """
+        data : dict
+            keys of ['train', 'test'] with values suitable for model.fit
+        """
+        self.params = params
+        self.model = model
+        self.R = R
+        self.eta = eta
+        self.best_val_loss = np.inf
```

Comment:
And if we're following the scikit-learn style (which I think we should), "learned" attributes like this would have a trailing underscore, and wouldn't exist till `.fit` is called.